File: java-module/core/src/main/java/com/enterprise/qa/core/ai/selfhealing/SelfHealingLocator.java
package com.enterprise.qa.core.ai.selfhealing;

import com.enterprise.qa.core.config.ConfigManager;
import lombok.extern.slf4j.Slf4j;
import org.openqa.selenium.By;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.StaleElementReferenceException;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * AI-powered self-healing locator that automatically recovers from broken selectors.
 * Uses multiple fallback strategies and scoring to find the best alternative element.
 */
@Slf4j
public class SelfHealingLocator {

    private final WebDriver driver;
    private final LocatorScorer scorer;
    private final Map<String, ElementSnapshot> snapshotCache;
    private final boolean enabled;

    public SelfHealingLocator(WebDriver driver) {
        this.driver = driver;
        this.scorer = new LocatorScorer();
        this.snapshotCache = new HashMap<>();
        this.enabled = ConfigManager.getInstance().isSelfHealingEnabled();
    }

    /**
     * Finds an element with self-healing capability.
     *
     * @param primaryLocator the primary locator to try first
     * @param elementName    a descriptive name for the element (used for healing)
     * @return the found element
     * @throws NoSuchElementException if element cannot be found even with healing
     */
    public WebElement findElement(By primaryLocator, String elementName) {
        // Try primary locator first
        try {
            WebElement element = driver.findElement(primaryLocator);
            if (element.isDisplayed()) {
                // Cache snapshot for future healing
                cacheElementSnapshot(elementName, element, primaryLocator);
                log.debug("Found element '{}' with primary locator", elementName);
                return element;
            }
        } catch (NoSuchElementException | StaleElementReferenceException e) {
            log.warn("Primary locator failed for '{}': {}", elementName, primaryLocator);
        }

        // If self-healing is disabled, throw immediately
        if (!enabled) {
            throw new NoSuchElementException("Element not found: " + elementName);
        }

        // Attempt self-healing
        return healAndFind(primaryLocator, elementName);
    }

    /**
     * Finds all elements matching a locator with self-healing capability.
     *
     * @param primaryLocator the primary locator to try
     * @param elementName    a descriptive name for the elements
     * @return list of found elements
     */
    public List<WebElement> findElements(By primaryLocator, String elementName) {
        try {
            List<WebElement> elements = driver.findElements(primaryLocator);
            if (!elements.isEmpty()) {
                return elements;
            }
        } catch (Exception e) {
            log.warn("Primary locator failed for '{}': {}", elementName, e.getMessage());
        }

        if (!enabled) {
            return Collections.emptyList();
        }

        // Try alternative strategies
        return healAndFindMultiple(primaryLocator, elementName);
    }

    /**
     * Attempts to heal a broken locator and find the element.
     *
     * @param originalLocator the original broken locator
     * @param elementName     the element name
     * @return the healed element
     */
    private WebElement healAndFind(By originalLocator, String elementName) {
        log.info("Attempting self-healing for element: {}", elementName);

        // Get cached snapshot if available
        ElementSnapshot snapshot = snapshotCache.get(elementName);

        // Generate alternative locators
        List<By> alternatives = generateAlternativeLocators(originalLocator, snapshot);

        // Try each alternative and score the results
        List<ScoredElement> candidates = new ArrayList<>();

        for (By alternative : alternatives) {
            try {
                List<WebElement> elements = driver.findElements(alternative);
                for (WebElement element : elements) {
                    if (element.isDisplayed()) {
                        double score = scorer.scoreElement(element, snapshot);
                        candidates.add(new ScoredElement(element, alternative, score));
                    }
                }
            } catch (Exception e) {
                log.trace("Alternative locator failed: {}", alternative);
            }
        }

        if (candidates.isEmpty()) {
            log.error("Self-healing failed for element: {}", elementName);
            throw new NoSuchElementException("Could not heal locator for: " + elementName);
        }

        // Sort by score and return the best match
        candidates.sort((a, b) -> Double.compare(b.score, a.score));
        ScoredElement best = candidates.get(0);

        log.info("Self-healed element '{}' with score {} using: {}",
                elementName, best.score, best.locator);

        // Cache the healed snapshot
        cacheElementSnapshot(elementName, best.element, best.locator);

        return best.element;
    }

    /**
     * Attempts to heal and find multiple elements.
     *
     * @param originalLocator the original locator
     * @param elementName     the element name
     * @return list of found elements
     */
    private List<WebElement> healAndFindMultiple(By originalLocator, String elementName) {
        log.info("Attempting self-healing for elements: {}", elementName);

        List<By> alternatives = generateAlternativeLocators(originalLocator, null);

        for (By alternative : alternatives) {
            try {
                List<WebElement> elements = driver.findElements(alternative);
                if (!elements.isEmpty()) {
                    log.info("Self-healed elements '{}' using: {}", elementName, alternative);
                    return elements;
                }
            } catch (Exception e) {
                log.trace("Alternative locator failed: {}", alternative);
            }
        }

        return Collections.emptyList();
    }

    /**
     * Generates alternative locators based on the original and cached snapshot.
     *
     * @param original the original locator
     * @param snapshot the cached element snapshot
     * @return list of alternative locators
     */
    private List<By> generateAlternativeLocators(By original, ElementSnapshot snapshot) {
        List<By> alternatives = new ArrayList<>();

        // Extract locator info from the original
        String locatorString = original.toString();

        // If we have a snapshot, use its attributes
        if (snapshot != null) {
            // Try by ID
            if (snapshot.getId() != null && !snapshot.getId().isEmpty()) {
                alternatives.add(By.id(snapshot.getId()));
            }

            // Try by name
            if (snapshot.getName() != null && !snapshot.getName().isEmpty()) {
                alternatives.add(By.name(snapshot.getName()));
            }

            // Try by class name
            if (snapshot.getClassName() != null && !snapshot.getClassName().isEmpty()) {
                String[] classes = snapshot.getClassName().split("\\s+");
                for (String cls : classes) {
                    if (!cls.isEmpty() && !isGenericClass(cls)) {
                        alternatives.add(By.className(cls));
                    }
                }
            }

            // Try by text content
            if (snapshot.getText() != null && !snapshot.getText().isEmpty()) {
                String text = snapshot.getText().trim();
                if (text.length() <= 50) {
                    alternatives.add(By.xpath("//*[normalize-space(text())='" + escapeXPath(text) + "']"));
                    alternatives.add(By.xpath("//*[contains(text(),'" + escapeXPath(text) + "')]"));
                }
            }

            // Try by data attributes
            for (Map.Entry<String, String> attr : snapshot.getDataAttributes().entrySet()) {
                alternatives.add(By.cssSelector("[" + attr.getKey() + "='" + attr.getValue() + "']"));
            }

            // Try by aria-label
            if (snapshot.getAriaLabel() != null && !snapshot.getAriaLabel().isEmpty()) {
                alternatives.add(By.cssSelector("[aria-label='" + snapshot.getAriaLabel() + "']"));
            }

            // Try by tag name with specific attributes
            String tag = snapshot.getTagName();
            if (tag != null) {
                if (snapshot.getType() != null) {
                    alternatives.add(By.cssSelector(tag + "[type='" + snapshot.getType() + "']"));
                }
                if (snapshot.getPlaceholder() != null) {
                    alternatives.add(By.cssSelector(tag + "[placeholder='" + snapshot.getPlaceholder() + "']"));
                }
            }
        }

        // Generate generic alternatives from original locator
        alternatives.addAll(generateGenericAlternatives(locatorString));

        return alternatives.stream()
                .filter(Objects::nonNull)
                .distinct()
                .collect(Collectors.toList());
    }

    /**
     * Generates generic alternatives from the original locator string.
     */
    private List<By> generateGenericAlternatives(String locatorString) {
        List<By> alternatives = new ArrayList<>();

        // Try to extract value from locator string
        if (locatorString.contains("By.id:")) {
            String id = extractValue(locatorString, "By.id:");
            // Try partial ID match
            alternatives.add(By.cssSelector("[id*='" + id + "']"));
            alternatives.add(By.cssSelector("[id$='" + id + "']"));
        }

        if (locatorString.contains("By.name:")) {
            String name = extractValue(locatorString, "By.name:");
            alternatives.add(By.cssSelector("[name*='" + name + "']"));
        }

        if (locatorString.contains("By.className:")) {
            String className = extractValue(locatorString, "By.className:");
            alternatives.add(By.cssSelector("[class*='" + className + "']"));
        }

        return alternatives;
    }

    /**
     * Caches an element snapshot for future healing attempts.
     */
    private void cacheElementSnapshot(String name, WebElement element, By locator) {
        try {
            ElementSnapshot snapshot = ElementSnapshot.capture(element);
            snapshot.setLastKnownLocator(locator.toString());
            snapshotCache.put(name, snapshot);
        } catch (Exception e) {
            log.debug("Failed to cache element snapshot: {}", e.getMessage());
        }
    }

    /**
     * Extracts a value from a locator string.
     */
    private String extractValue(String locatorString, String prefix) {
        int start = locatorString.indexOf(prefix) + prefix.length();
        return locatorString.substring(start).trim();
    }

    /**
     * Escapes special characters for XPath.
     */
    private String escapeXPath(String text) {
        if (text.contains("'")) {
            return "concat('" + text.replace("'", "',\"'\",'") + "')";
        }
        return text;
    }

    /**
     * Checks if a class name is generic and not useful for locating.
     */
    private boolean isGenericClass(String className) {
        Set<String> genericClasses = Set.of(
                "container", "wrapper", "content", "row", "col", "btn",
                "active", "disabled", "hidden", "visible", "clearfix"
        );
        return genericClasses.contains(className.toLowerCase());
    }

    /**
     * Clears the element snapshot cache.
     */
    public void clearCache() {
        snapshotCache.clear();
    }

    /**
     * Gets the number of cached element snapshots.
     */
    public int getCacheSize() {
        return snapshotCache.size();
    }

    /**
     * Inner class to hold scored element candidates.
     */
    private static class ScoredElement {
        final WebElement element;
        final By locator;
        final double score;

        ScoredElement(WebElement element, By locator, double score) {
            this.element = element;
            this.locator = locator;
            this.score = score;
        }
    }
}

File: js-module/playwright/src/ai/SelfHealingLocator.ts
import { Page, Locator } from '@playwright/test';

/**
 * Element snapshot for self-healing.
 */
interface ElementSnapshot {
  selector: string;
  tagName: string;
  id?: string;
  name?: string;
  className?: string;
  text?: string;
  type?: string;
  placeholder?: string;
  ariaLabel?: string;
  dataTestId?: string;
  href?: string;
  x: number;
  y: number;
  width: number;
  height: number;
  capturedAt: number;
}

/**
 * Self-healing locator that automatically recovers from broken selectors.
 * Uses multiple fallback strategies to find elements when primary locators fail.
 */
export class SelfHealingLocator {
  private readonly page: Page;
  private readonly snapshotCache: Map<string, ElementSnapshot>;
  private readonly enabled: boolean;

  constructor(page: Page, enabled = true) {
    this.page = page;
    this.snapshotCache = new Map();
    this.enabled = enabled;
  }

  /**
   * Finds an element with self-healing capability.
   * @param selector - The primary selector
   * @param elementName - A descriptive name for the element
   * @returns The found locator
   */
  async find(selector: string, elementName: string): Promise<Locator> {
    // Try primary selector first
    const primaryLocator = this.page.locator(selector);

    try {
      if ((await primaryLocator.count()) > 0 && (await primaryLocator.first().isVisible())) {
        // Cache snapshot for future healing
        await this.cacheSnapshot(elementName, primaryLocator.first(), selector);
        return primaryLocator.first();
      }
    } catch (e) {
      console.warn(`Primary selector failed for '${elementName}': ${selector}`);
    }

    if (!this.enabled) {
      throw new Error(`Element not found: ${elementName}`);
    }

    // Attempt self-healing
    return await this.healAndFind(selector, elementName);
  }

  /**
   * Attempts to heal a broken locator and find the element.
   */
  private async healAndFind(originalSelector: string, elementName: string): Promise<Locator> {
    console.log(`Attempting self-healing for element: ${elementName}`);

    const snapshot = this.snapshotCache.get(elementName);
    const alternatives = this.generateAlternatives(originalSelector, snapshot);

    for (const alternative of alternatives) {
      try {
        const locator = this.page.locator(alternative);
        const count = await locator.count();

        if (count > 0) {
          const element = locator.first();
          if (await element.isVisible()) {
            const score = await this.scoreElement(element, snapshot);
            console.log(`Self-healed '${elementName}' using: ${alternative} (score: ${score})`);

            // Cache the healed selector
            await this.cacheSnapshot(elementName, element, alternative);
            return element;
          }
        }
      } catch (e) {
        // Try next alternative
      }
    }

    throw new Error(`Could not heal locator for: ${elementName}`);
  }

  /**
   * Generates alternative selectors.
   */
  private generateAlternatives(originalSelector: string, snapshot?: ElementSnapshot): string[] {
    const alternatives: string[] = [];

    if (snapshot) {
      // Try by ID
      if (snapshot.id) {
        alternatives.push(`#${snapshot.id}`);
        alternatives.push(`[id="${snapshot.id}"]`);
        alternatives.push(`[id*="${snapshot.id}"]`);
      }

      // Try by data-testid
      if (snapshot.dataTestId) {
        alternatives.push(`[data-testid="${snapshot.dataTestId}"]`);
        alternatives.push(`[data-test-id="${snapshot.dataTestId}"]`);
        alternatives.push(`[data-cy="${snapshot.dataTestId}"]`);
      }

      // Try by name
      if (snapshot.name) {
        alternatives.push(`[name="${snapshot.name}"]`);
      }

      // Try by aria-label
      if (snapshot.ariaLabel) {
        alternatives.push(`[aria-label="${snapshot.ariaLabel}"]`);
      }

      // Try by placeholder
      if (snapshot.placeholder) {
        alternatives.push(`[placeholder="${snapshot.placeholder}"]`);
      }

      // Try by text content
      if (snapshot.text && snapshot.text.length <= 50) {
        alternatives.push(`text="${snapshot.text}"`);
        alternatives.push(`text=${snapshot.text}`);
        alternatives.push(`*:has-text("${snapshot.text}")`);
      }

      // Try by tag and type
      if (snapshot.tagName && snapshot.type) {
        alternatives.push(`${snapshot.tagName}[type="${snapshot.type}"]`);
      }

      // Try by class
      if (snapshot.className) {
        const classes = snapshot.className.split(/\s+/).filter((c) => !this.isGenericClass(c));
        for (const cls of classes) {
          alternatives.push(`.${cls}`);
        }
      }
    }

    // Generate alternatives from original selector
    alternatives.push(...this.generateFromOriginal(originalSelector));

    return [...new Set(alternatives)]; // Remove duplicates
  }

  /**
   * Generates alternatives from the original selector.
   */
  private generateFromOriginal(selector: string): string[] {
    const alternatives: string[] = [];

    // Try partial ID match
    const idMatch = selector.match(/#([a-zA-Z0-9_-]+)/);
    if (idMatch) {
      alternatives.push(`[id*="${idMatch[1]}"]`);
      alternatives.push(`[id$="${idMatch[1]}"]`);
    }

    // Try partial class match
    const classMatch = selector.match(/\.([a-zA-Z0-9_-]+)/);
    if (classMatch) {
      alternatives.push(`[class*="${classMatch[1]}"]`);
    }

    return alternatives;
  }

  /**
   * Scores how well an element matches a snapshot.
   */
  private async scoreElement(locator: Locator, snapshot?: ElementSnapshot): Promise<number> {
    if (!snapshot) {
      return 0.5; // Base score without snapshot
    }

    let score = 0;
    let weights = 0;

    try {
      const element = await locator.elementHandle();
      if (!element) return 0;

      // Tag name
      const tagName = await locator.evaluate((el) => el.tagName.toLowerCase());
      if (tagName === snapshot.tagName?.toLowerCase()) {
        score += 0.15;
      }
      weights += 0.15;

      // ID
      const id = await locator.getAttribute('id');
      if (id && id === snapshot.id) {
        score += 0.2;
      }
      weights += 0.2;

      // Text content
      const text = await locator.textContent();
      if (text && snapshot.text) {
        const similarity = this.stringSimilarity(text.trim(), snapshot.text);
        score += 0.15 * similarity;
      }
      weights += 0.15;

      // Position
      const box = await locator.boundingBox();
      if (box && snapshot.x !== undefined) {
        const dx = Math.abs(box.x - snapshot.x);
        const dy = Math.abs(box.y - snapshot.y);
        const distance = Math.sqrt(dx * dx + dy * dy);
        const positionScore = Math.max(0, 1 - distance / 200);
        score += 0.1 * positionScore;
      }
      weights += 0.1;

      // Size
      if (box && snapshot.width && snapshot.height) {
        const widthRatio =
          Math.min(box.width, snapshot.width) / Math.max(box.width, snapshot.width);
        const heightRatio =
          Math.min(box.height, snapshot.height) / Math.max(box.height, snapshot.height);
        score += 0.1 * ((widthRatio + heightRatio) / 2);
      }
      weights += 0.1;

      return weights > 0 ? score / weights : 0;
    } catch (e) {
      return 0;
    }
  }

  /**
   * Caches an element snapshot.
   */
  private async cacheSnapshot(name: string, locator: Locator, selector: string): Promise<void> {
    try {
      const snapshot: ElementSnapshot = {
        selector,
        tagName: await locator.evaluate((el) => el.tagName.toLowerCase()),
        id: (await locator.getAttribute('id')) || undefined,
        name: (await locator.getAttribute('name')) || undefined,
        className: (await locator.getAttribute('class')) || undefined,
        text: (await locator.textContent())?.trim().substring(0, 100) || undefined,
        type: (await locator.getAttribute('type')) || undefined,
        placeholder: (await locator.getAttribute('placeholder')) || undefined,
        ariaLabel: (await locator.getAttribute('aria-label')) || undefined,
        dataTestId:
          (await locator.getAttribute('data-testid')) ||
          (await locator.getAttribute('data-test-id')) ||
          undefined,
        href: (await locator.getAttribute('href')) || undefined,
        x: 0,
        y: 0,
        width: 0,
        height: 0,
        capturedAt: Date.now(),
      };

      const box = await locator.boundingBox();
      if (box) {
        snapshot.x = box.x;
        snapshot.y = box.y;
        snapshot.width = box.width;
        snapshot.height = box.height;
      }

      this.snapshotCache.set(name, snapshot);
    } catch (e) {
      console.debug(`Failed to cache snapshot for ${name}: ${e}`);
    }
  }

  /**
   * Calculates string similarity (Levenshtein-based).
   */
  private stringSimilarity(s1: string, s2: string): number {
    if (s1 === s2) return 1;
    if (!s1 || !s2) return 0;

    const maxLen = Math.max(s1.length, s2.length);
    if (maxLen === 0) return 1;

    const distance = this.levenshteinDistance(s1, s2);
    return 1 - distance / maxLen;
  }

  /**
   * Calculates Levenshtein distance.
   */
  private levenshteinDistance(s1: string, s2: string): number {
    const m = s1.length;
    const n = s2.length;
    const dp: number[][] = Array(m + 1)
      .fill(null)
      .map(() => Array(n + 1).fill(0));

    for (let i = 0; i <= m; i++) dp[i][0] = i;
    for (let j = 0; j <= n; j++) dp[0][j] = j;

    for (let i = 1; i <= m; i++) {
      for (let j = 1; j <= n; j++) {
        const cost = s1[i - 1] === s2[j - 1] ? 0 : 1;
        dp[i][j] = Math.min(dp[i - 1][j] + 1, dp[i][j - 1] + 1, dp[i - 1][j - 1] + cost);
      }
    }

    return dp[m][n];
  }

  /**
   * Checks if a class name is generic.
   */
  private isGenericClass(className: string): boolean {
    const genericClasses = new Set([
      'container',
      'wrapper',
      'content',
      'row',
      'col',
      'btn',
      'active',
      'disabled',
      'hidden',
      'visible',
      'clearfix',
    ]);
    return genericClasses.has(className.toLowerCase());
  }

  /**
   * Clears the snapshot cache.
   */
  clearCache(): void {
    this.snapshotCache.clear();
  }

  /**
   * Gets the cache size.
   */
  getCacheSize(): number {
    return this.snapshotCache.size;
  }
}
